### PR TITLE
Create test database / API

### PR DESCRIPTION
**Pull request by sidiousvic**
Creates a test database class to mock the TypeORM models as closely as possible to production. 

```js
export class TestDatabase {...}
```

Uses `supertest` as a test client for API calls.
```js
  public get app(): supertest.SuperTest<supertest.Test> {
    return supertest(this._app)
```

Developers now have the ability to test requests from the client layer as follows.

```js
 describe('GET /users/', () => {
    it('retrieves a User', async () => {
      const response = await database.app
        .get('/api/v1/users')
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.statusCode).toBe(200)
    })
  })
```

# Notes

- [x] Might want to reorganize model tests so they live closer to the model itself.
- [x] Added a `test:watch` command which is useful for continually running test specs.
- [x] It will be a good idea to practice a TDD cycle (write a failing test, implement the pass) with this setup enabled.
- [x] ✸ For example, creating tests for the auth token logic might be good to simulate usage before implementation.
- [x] Keep up the hard work @Joboa.

### CODE REVIEW

File: backend/jest.config.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { pathsToModuleNameMapper } from 'ts-jest/utils'
+import { compilerOptions } from './tsconfig.json'
+
+module.exports = {
+  preset: 'ts-jest',
+}
```
Comment (sidiousvic) [inline backend/jest.config.ts:6]: Turned this into a TS module to accommodate typing issues.

File: backend/src/models/__test__/TestDatabase.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  private connectionOptions: ConnectionOptions = {
+    type: 'postgres',
+    port: parseInt(process.env.PORT_DB),
+    username: process.env.USERNAME_DB,
+    password: process.env.PASSWORD_DB,
+    database: 'expense_manager_test',
+    entities: [User, Expense],
+    synchronize: true,
+    logging: false,
+  }
```
Comment (sidiousvic) [inline backend/src/models/__test__/TestDatabase.ts:32]: @Joboa make sure to create an `expense_manager` (production) and `expense_manager_test` locally in Postgres.

```
$ psql
#= CREATE DATABASE expense_manager;
#= CREATE DATABASE expense_manager_test;
```

File: backend/src/models/__test__/user.spec.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  user.lastname = 'Sidious'
+  user.password = 'pieTillIDie'
+
+  beforeEach(async () => {
+    await database.init()
+  })
+
+  afterEach(async () => {
+    await database.kill()
+  })
```
Comment (sidiousvic) [inline backend/src/models/__test__/user.spec.ts:20]: Database/server connection should be cleaned up, don't forget about this when testing.

File: backend/src/models/__test__/user.spec.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import { clone } from '../../utils/clone'
+
+describe('Testing user component', () => {
+  const database = new TestDatabase()
+  const user = clone(new User())
+
+  user.email = 'x@email.com'
+  user.firstname = 'Vic'
+  user.lastname = 'Sidious'
+  user.password = 'pieTillIDie'
```
Comment (sidiousvic) [inline backend/src/models/__test__/user.spec.ts:12]: This is for subsequent tests, for when you need to test a `POST` request perhaps, you will need a mock user.

File: backend/jest.config.ts (hunk 1 above)
Comment (Joboa) [reply backend/jest.config.ts:6]: Okay

File: backend/src/models/__test__/TestDatabase.ts (hunk 2 above)
Comment (Joboa) [reply backend/src/models/__test__/TestDatabase.ts:32]: Okay, I will do that.

File: backend/src/models/__test__/user.spec.ts (hunk 3 above)
Comment (Joboa) [reply backend/src/models/__test__/user.spec.ts:20]: Well noted

File: backend/src/models/__test__/user.spec.ts (hunk 4 above)
Comment (Joboa) [reply backend/src/models/__test__/user.spec.ts:12]: Okay

**Review by Joboa — COMMENTED** [review]
I have gone through all the changes. I have learn something new today. I am amazed. In case of any question, I will message  you on Discord. Thank you.

**Review by Joboa — COMMENTED** [review]
`export function clone<T>(instance: T): T {
  const copy = new (instance.constructor as { new (): T })()
  Object.assign(copy, instance)
  return copy
}`

Please I really don't get the second line of this code:  `const copy = new (instance.constructor as { new (): T })()`. I would like you to explain more on this code.

**Comment by sidiousvic** [comment]
@Joboa First one note—feel free to use ``` triple backticks to format and highlight your code, as such: 

```
```ts
```

```ts
export function clone<T>(instance: T): T { 
  const copy = new (instance.constructor as { new (): T })() 
  Object.assign(copy, instance) 
  return copy 
}
```

> Please I really don't the second line of this code: 

```ts
const copy = new (instance.constructor as { new (): T })()`. 
```
This function will take an instance of a class `instance.constructor()`, and use its constructor to create a new instance with all of the properties of the original. The part with ``` { new (): T }``` is just a type literal for a constructor function.

This might make it clearer:

```ts
type Constructor = { new (): T };

const copy = new (instance.constructor as Constructor)()`.
```

Or in regular JS,

```js
const copy = new instance.constructor()
```

 I wanted this so I could extend class instances in tests without mutating them.

**Pull request merged**